runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ReadTranslation.java

```diff
@@ -152,6 +153,19 @@ private static SdkFunctionSpec toProto(UnboundedSource<?, ?> source, SdkComponen
     }
   }
 
+  public static IsBounded.Enum isBounded(AppliedPTransform<?, ?, ?> transform) {
```

Comment:
@tgroh Is this a good place to put such a method or is there already an easier method for determining whether a PTransform/PCollection is unbounded?

Reply:
I think this is precisely `sourceIsBounded` for all well formed `ReadPayload`s (one method up)

Reply:
yep, I thought that as well but it takes a `ReadPayload` and when traversing the pipeline we don't know which transform is a read.

Another alternative would be 
```
try {
  rawSource = ReadTranslation.unboundedSourceFromTransform(
      (AppliedPTransform<PBegin, PCollection<T>, PTransform<PBegin, PCollection<T>>>)
          node.toAppliedPTransform(getPipeline()));
  // we have an unbounded source
} catch (IOException e) {
  // no unbounded source
}
```

but I'm not sure I like the exceptions-as-control-flow thing.

Reply:
Ah; I see. If we don't know that it's a read transform I'm generically opposed to routing it through `ReadTranslation`, plus I also don't really like the use of exceptions as control flow.

Each `PCollection` contains its boundedness, so you can just inspect the outputs of the transform (which are labelled as `PValues`, but I believe are all `PCollections` at this point; maybe worth the check regardless). They have the `PCollection.isBounded()` method, in both proto and java forms